### PR TITLE
Adding a sample to show how to use, .NET Framework 4.7.2 (and .NET core), the TokenCache providers offered by Microsoft.Identity.Web.

### PR DESCRIPTION
**Pull request by jmprieur**
## Purpose
This sample aims at helping the migration from ADAL to MSAL of developers who cannot use ASP.NET Core (otherwise they would use the full potential of [Microsoft.Identity.Web](https://github.com/AzureAD/microsoft-identity-web/wiki/1.0.0))

Solves https://github.com/AzureAD/microsoft-identity-web/issues/765

## Does this introduce a breaking change?

```
[ ] Yes
[ x] No
```

## Pull Request Type
New sample

## How to Test
*  Get the code

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
```

* Test the code
Contact jmprieur for the client secret

## Other Information
I have not really added services for Cosmos, Redis and Sql. Would need to be done. This also provides a nice end to end test.

### CODE REVIEW

File: ConfidentialClientTokenCache/Program.cs [hunk 1]
```diff
@@ -0,0 +1,189 @@
+﻿using Microsoft.Azure.Cosmos.Fluent;
```
Comment (jennyf19) [inline ConfidentialClientTokenCache/Program.cs]:
```suggestion
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

﻿using Microsoft.Azure.Cosmos.Fluent;
```

File: ConfidentialClientTokenCache/Program.cs [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+
+            // Acquire a token (twice)
+            var result = await app.AcquireTokenForClient(scopes)
+                .ExecuteAsync();
+            Console.WriteLine(result.AuthenticationResultMetadata.TokenSource);
+
+            result = await app.AcquireTokenForClient(scopes)
+                .ExecuteAsync();
+            Console.WriteLine(result.AuthenticationResultMetadata.TokenSource);
+
```
Comment (jennyf19) [inline ConfidentialClientTokenCache/Program.cs]:
```suggestion
```

File: ConfidentialClientTokenCache/Program.cs [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            {
+                throw new ArgumentNullException(nameof(services));
+            }
+
+            services.AddDistributedMemoryCache();
+            services.AddSingleton<IMsalTokenCacheProvider, MsalDistributedTokenCacheAdapter>();
+            return services;
+        }
+    }
+
```
Comment (jennyf19) [inline ConfidentialClientTokenCache/Program.cs:167]:
```suggestion
```

File: ConfidentialClientTokenCache/Program.cs [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+            Console.WriteLine(result.AuthenticationResultMetadata.TokenSource);
+
+        }
+
+        /// <summary>
+        /// Creates a token cache (implementation of your choice)
+        /// </summary>
+        /// <param name="cacheImplementation">implementation for the token cache</param>
+        /// <returns>An Msal Token cache provider</returns>
+        private static IMsalTokenCacheProvider CreateTokenCache(CacheImplementationDemo cacheImplementation=CacheImplementationDemo.InMemory)
```
Comment (jennyf19) [inline ConfidentialClientTokenCache/Program.cs:75]: wonder how much of this can be moved into one of the libraries? make it even simpler.

Comment (jmprieur) [reply ConfidentialClientTokenCache/Program.cs:75]: Awesome idea. Let's think what we can do for the options.
This requires, however to reference NuGet packages that customers might not want
Also the CosmosDb token distributed cache provider is still in preview.

**Review by jennyf19 — APPROVED** [review]

**Review by neha-bhargava — APPROVED** [review]
Looks good. Thanks!

**Review by trwalke — APPROVED** [review]
Thanks JM

File: ConfidentialClientTokenCache/Program.cs [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+                    break;
+
+                case CacheImplementationDemo.StackExchangeRedis:
+                    // Redis token cache
+                    // Requires to reference Microsoft.Extensions.Caching.StackExchangeRedis
+                    services.AddDistributedTokenCaches();
+                    services.AddStackExchangeRedisCache(options =>
+                    {
+                        options.Configuration = "localhost";
+                        options.InstanceName = "SampleInstance";
```
Comment (henrik-me) [inline ConfidentialClientTokenCache/Program.cs:116]: should these also be injected from the Configuration (same above) - and having the config object initialized with these values?

File: ConfidentialClientTokenCache/Program.cs [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+    /// in confidential client applications both in .NET Core, or .NET Framework 4.7.2
+    /// Note that if you write a .NET Core web app or web API, Microsoft recommends that you use the
+    /// IDownstreamApi or ITokenAcquisition interfaces directly (instead of MSAL)
+    /// </summary>
+    static class Program
+    {
+        static async Task Main(string[] args)
+        {
+            string clientId = "812287fd-3ea3-49c6-b4ab-e8d41dea1f17";
+            string clientSecret = "[Enter here the tenantID or domain name for your Azure AD tenant]";
```
Comment (henrik-me) [inline ConfidentialClientTokenCache/Program.cs:41]: >Enter here the tenantID or domain name for your Azure AD tenant] [](start = 36, length = 64)

should say enter secret ?

**Review by henrik-me — APPROVED** [review]
:shipit:

File: ConfidentialClientTokenCache/Program.cs [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+            }
+
+            IServiceProvider serviceProvider = services.BuildServiceProvider();
+            IMsalTokenCacheProvider msalTokenCacheProvider = serviceProvider.GetRequiredService<IMsalTokenCacheProvider>();
+            return msalTokenCacheProvider;
+        }
+    }
+
+    #region Will be brought by Microsoft.Identity.Web in 1.4.0
+    // See https://github.com/AzureAD/microsoft-identity-web/pull/777
```
Comment (henrik-me) [inline ConfidentialClientTokenCache/Program.cs:143]: nice !

File: ConfidentialClientTokenCache/Program.cs (hunk 5 above)
Comment (jmprieur) [reply ConfidentialClientTokenCache/Program.cs:116]: I've added an issue to do it later: https://github.com/Azure-Samples/active-directory-dotnet-v1-to-v2/issues/20
proposing to postpone it after the release (it's more urgent / important to release MSAL.NET)

**Pull request merged**